proxy/grpcproxy/cluster.go

```diff
@@ -15,20 +15,99 @@
 package grpcproxy
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/naming"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 
 	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	gnaming "google.golang.org/grpc/naming"
 )
 
 type clusterProxy struct {
 	client *clientv3.Client
+	wa     gnaming.Watcher
+	donec  <-chan struct{}
+	umu    sync.RWMutex
+	umap   map[string]struct{}
 }
 
-func NewClusterProxy(c *clientv3.Client) pb.ClusterServer {
-	return &clusterProxy{
+// NewClusterProxy takes optional prefix to fetch grpc-proxy member endpoints.
+// The returned channel is closed when there is grpc-proxy endpoint registered
```

Comment:
This isn't a good interface; the return results are too dependent on the arguments. Just have `donec` closed when all goroutines associated with the cluster proxy are shutdown (like the other proxy services). Right now it will need special code for `donec` based on whether a prefix was given or not.